rpc_new/provider/smoldot.ts

```diff
@@ -0,0 +1,92 @@
+import { start } from "../../deps/smoldot.ts";
+import { Chain, Client, ClientOptions } from "../../deps/smoldot/client.d.ts";
+import * as msg from "../messages.ts";
+import { nextIdFactory, Provider, ProviderConnection, ProviderListener } from "./base.ts";
+import { ProviderCloseError, ProviderHandlerError, ProviderSendError } from "./errors.ts";
+
+// TODO
+type SmoldotSendErrorData = any;
+type SmoldotHandlerErrorData = any;
+type SmoldotCloseErrorData = any;
+
+let client: undefined | Client;
+const connections = new Map<string, SmoldotProviderConnection>();
+class SmoldotProviderConnection
+  extends ProviderConnection<Chain, SmoldotSendErrorData, SmoldotHandlerErrorData>
+{}
+
+const nextId = nextIdFactory();
+
+export const smoldotProvider: Provider<
+  string,
+  SmoldotSendErrorData,
+  SmoldotHandlerErrorData,
+  SmoldotCloseErrorData
+> = (chainSpec, listener) => {
+  return {
+    nextId,
+    send: (message) => {
+      (async () => {
+        const conn = await connection(chainSpec, listener);
+        try {
+          return conn.inner.sendJsonRpc(JSON.stringify(message));
+        } catch (e) {
+          return new ProviderSendError(e);
+        }
+      })();
+    },
+    release: async () => {
+      const { cleanUp, listeners, inner } = await connection(chainSpec, listener);
+      listeners.delete(listener);
+      if (!listeners.size) {
+        connections.delete(chainSpec);
+        cleanUp();
+        try {
+          return inner.remove();
+        } catch (e) {
+          return new ProviderCloseError(e);
+        }
+      }
+    },
+  };
+};
+
+async function connection(
+  chainSpec: string,
+  listener: ProviderListener<SmoldotSendErrorData, SmoldotHandlerErrorData>,
+): Promise<SmoldotProviderConnection> {
+  if (!client) {
+    client = start(
+      <ClientOptions> {
+        forbidTcp: true,
+        forbidNonLocalWs: true,
+      },
+    );
+  }
+  let conn = connections.get(chainSpec);
+  if (!conn) {
+    // TODO: try catch this and send through handler within a `ProviderHandlerError`
+    const inner = await client.addChain({ chainSpec });
+    conn = new SmoldotProviderConnection(inner, () => {
+      try {
+        inner.remove();
+      } catch (_e) { /* TODO */ }
+    });
+    connections.set(chainSpec, conn);
+    const loop = async () => {
+      try {
+        conn!.forEachListener(msg.parse(await inner.nextJsonRpcResponse()));
+      } catch (e) {
+        conn!.forEachListener(new ProviderHandlerError(e));
+        await loop();
+      }
```

Comment:
Looks like right now each `loop` will only ever output one response?

```suggestion
      while (true) {
        try {
          conn!.forEachListener(msg.parse(await inner.nextJsonRpcResponse()));
        } catch (e) {
          conn!.forEachListener(new ProviderHandlerError(e));
        }
      }
```

Reply:
How does one break out of this? Let's tackle this provider more in depth in a subsequent PR. This is haphazard atm.